promql/parser/prettier.go

```diff
@@ -0,0 +1,579 @@
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package parser
+
+import (
+	"regexp"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+const (
+	// itemNotFound is used to initialize the index of an item.
+	itemNotFound  = -1
+	columnLimit   = 100
+	defaultIndent = "  "
+)
+
+// Prettier handles the prettifying and formatting operation over a
+// list of rules files or a single expression.
+type Prettier struct {
+	Node       Expr
+	expression string
+	pd         *padder
+}
+
+type padder struct {
+	indent      string
+	columnLimit int
+	// isNewLineApplied represents a new line was added previously, after
+	// appending the item to the result. It should be set to default
+	// after applying baseIndent. This property is dependent across the padder states
+	// and hence needs to be carried along while formatting.
+	isNewLineApplied bool
+	// isPreviousItemComment confirms whether the previous formatted item was
+	// a comment. This property is dependent on previous padder states and hence needs
+	// to be carried along while formatting.
+	isPreviousItemComment bool
+	// expectAggregationLabels expects the following item(s) to be labels of aggregation
+	// expression. Labels are used in AggregateExpr and in VectorSelector but only the ones
+	// in VectorSelector are splittable.
+	expectAggregationLabels bool
+	// labelsSplittable determines whether the current formatting region lies in labels
+	// matchers in VectorSelector. Since labels in VectorSelectors are only splittable,
+	// is property is applied only after encountering a Left_Brace. This is dependent across
+	// multiple padder states and hence needs to be carried along.
+	labelsSplittable bool
+	isFirstLabel     bool
+}
+
+// PromqlPrettier returns a new prettier.
+func PromqlPrettier(expression string) *Prettier {
+	return &Prettier{
+		expression: expression,
+		pd:         &padder{indent: defaultIndent, isNewLineApplied: true, columnLimit: columnLimit},
+	}
+}
+
+// Prettify prettifies the current expression in the prettier. It standardizes the input expression,
+// sorts the items slice and calls the prettify.
+func (p *Prettier) Prettify() (string, error) {
+	standardizeExprStr := stringifyItems(
+		p.sortItems(p.lexItems(p.expression)),
+	)
+	if err := p.parseExpr(standardizeExprStr); err != nil {
+		return p.expression, err
+	}
+	formattedExpr, err := p.prettify(p.lexItems(standardizeExprStr))
+	if err != nil {
+		return p.expression, errors.Wrap(err, "Prettify")
+	}
+	return formattedExpr, nil
+}
+
+func (p *Prettier) sortItems(items []Item) []Item {
```

Comment:
Can you describe what do you mean by sorting here? If it is the re-arrangement of lex items, then we might as well call it `rearrangeItems` (with comments ofcourse).